### DISCUSSION TITLE
IKE_DOWN, without IKE_UP

### DISCUSSION THREAD

**Discussion by amvarunn** [strongswan/strongswan#987] [Q&A]
We are running application which interacts with strongswan through vici interfaces. We allocate the memory IKE SA upon an IKE UP event, and deallocate upon IKE DOWN event coming up with same uniquid. We receive an IKE down event for the IKE initiation triggered by our vici call in case of tunnel creation failure, which is expected since the tunnel creation failed. However, strongswan seems to be re-initiating IKE SA with next uniquid (no vici call for tunnel initiation this time), for which we receive just IKE DOWN upon failure (without any corresponding IKE UP since the tunnel creation failed anyway).  Is this behaviour expected? This kind of impacts the memory allocation model that is described above since we seem to be receiving more IKE down events than IKE UPs.

```
    370 Mon Apr  4 08:53:32 2022 07[IKE] <IPVX_192-168-49-113_192-168-49-112_3869|25> IKE_SA deleted
    371 Mon Apr  4 08:54:08 2022 08[IKE] <IPVX_192-168-49-113_192-168-49-111_3867|26> initiating IKE_SA 192-168-49-113_192-168-49-111_3867[28] to 192.168.49.111
    372 Mon Apr  4 08:54:08 2022 11[CFG] <IPVX_192-168-49-113_192-168-49-111_3867|26> selected proposal: IKE:AES_CBC_128/HMAC_SHA2_2        56_128/PRF_AES128_XCBC/ECP_256
    373 Mon Apr  4 08:54:08 2022 11[IKE] <IPVX_192-168-49-113_192-168-49-111_3867|26> scheduling rekeying in 648s
    374 Mon Apr  4 08:54:08 2022 11[IKE] <IPVX_192-168-49-113_192-168-49-111_3867|26> scheduling reauthentication in 822s
    375 Mon Apr  4 08:54:08 2022 11[IKE] <IPVX_192-168-49-113_192-168-49-111_3867|26> maximum IKE_SA lifetime 738s
    376 Mon Apr  4 08:54:08 2022 11[IKE] <IPVX_192-168-49-113_192-168-49-111_3867|26>  IKE_SA 192-168-4 9-113_192-168-49-111_3867[28] rekeyed between 192.168.49.113[C=a, ST=b, L=c, O=d, OU=e, CN=f, E=g]...192.168.49.111[C=a, ST=b, L=c, O=d, OU=e, CN=f, E=g]
    377 Mon Apr  4 08:54:08 2022 11[IKE] <192-168-49-113_192-168-49-111_3867|26> rescheduling reauthentication in 204s after re        keying, lifetime reduced to 294s
    378 Mon Apr  4 08:54:08 2022 11[IKE] <IPVX_192-168-49-113_192-168-49-111_3867|26> deleting IKE_SA 192-168-49-113_192-168-49-111_3867[26] between 192.168.49.113[C=a, ST=b, L=c, O=d, OU=e, CN=f, E=g]...192.168.49.111[C=a, ST=b, L=c, O=d, OU=e, CN=f, E=g]
    379 Mon Apr  4 08:54:08 2022 11[IKE] <192-168-49-113_192-168-49-111_3867|26> sending DELETE for IKE_SA 192-168-49-113_192-168-49-111_3867[26]
    380 Mon Apr  4 08:54:08 2022 15[IKE] <192-168-49-113_192-168-49-111_3867|26> IKE_SA deleted
```

**Comment by tobiasbrunner** [comment]
> We receive an IKE down event for the IKE initiation triggered by our vici call in case of tunnel creation failure, which is expected since the tunnel creation failed.

I don't understand what you mean here. There is never a "down" event if connecting fails. Such events are only triggered once an IKE_SA has been established successfully (i.e. after an "up" event).

> However, strongswan seems to be re-initiating IKE SA with next uniquid

Triggered by what? `keyingtries`? `close|dpd_action`?

> (no vici call for tunnel initiation this time)

There is no event for this, whether it's the initial initiation or a later one, only once an IKE_SA is established will there be an "up" event.

> for which we receive just IKE DOWN upon failure (without any corresponding IKE UP since the tunnel creation failed anyway).

Again, there is generally no "down" event for unestablished IKE_SAs, so not sure what exactly you are referring to here.

> This kind of impacts the memory allocation model that is described above since we seem to be receiving more IKE down events than IKE UPs.

If you never allocated anything for a particular ID, why would a down event for it be a problem anyway?

Please provide detailed steps to reproduce this (or at least a more detailed/complete log that shows what's going on).

**Reply by amvarunn** [reply]
Thank you, I have some trouble recreating this issue. Will update this thread with the logs when I do.
